### PR TITLE
Bump PyO3 and rust-numpy to latest release

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

PyO3 and rust-numpy just released version 0.17.0 respectively (and both
are linked). This commit bumps the version used of both libraries to the
latest release. For the changelogs see:

### Details and comments

https://github.com/PyO3/pyo3/releases/tag/v0.17.0

and

https://github.com/PyO3/rust-numpy/releases/tag/v0.17.0

### CODE REVIEW

**Comment by mtreinish** [comment]
Looks like the conversion behavior changed in pyo3 0.17.0:

```
test.python.result.test_counts.TestCounts.test_marginal_distribution_numpy_indices
----------------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File "/tmp/terra-tests/test/python/result/test_counts.py", line 60, in test_marginal_distribution_numpy_indices
    result = utils.marginal_distribution(counts_obj, indices)

      File "/home/vsts/work/1/s/qiskit/result/utils.py", line 231, in marginal_distribution
    res = results_rs.marginal_counts(counts, indices)

    TypeError: argument 'indices': 'ndarray' object cannot be converted to 'Sequence'

```

looking at the migration guide I expect this is caused by the addition of the `isinstance` call on conversion: https://github.com/PyO3/pyo3/blob/v0.17.0/guide/src/migration.md#type-checks-have-been-changed-for-pymapping-and-pysequence-types as numpy arrays don't subclass `Sequence` and I don't think most C extension sequence types will either which will cause a problem. I'll report this upstrean

**Comment by mtreinish** [comment]
I've marked this as for 0.23.0 although I don't think there is actually a blocker for including it as part of 0.22.0. However, since the 0.22.0.rc1 is pending I say we merge this after `main` switches over to 0.23.0 and if we're happy with how it works on main we can consider backporting it to `stable/0.22` for inclusion in the final 0.22.0 release. I'm mildly concerned there might be some performance overhead with the new version of PyO3.

File: Cargo.toml [hunk 1]
```diff
@@ -11,7 +11,7 @@ crate-type = ["cdylib"]
 
 [dependencies]
 rayon = "1.5"
-numpy = "0.16.2"
+numpy = "0.17.0"
```
Comment (kevinhartman) [inline Cargo.toml:14]: Is there a reason to use `0.17.0` here instead of `0.17.2` like the other bump?

Comment (mtreinish) [reply Cargo.toml:14]: not particularly, the cargo.lock has 0.17.2 now as that gets pulled in by the this, but it probably makes sense to update this in lockstep to ensure we're getting all the bugfixes as the required version (not just relying on our lock version)

```suggestion
numpy = "0.17.2"
```

**Review by kevinhartman — APPROVED** [review]
LGTM!

**Pull request merged**